### PR TITLE
feat(client): add get_labels, get_edge_types, traverse

### LINKED ISSUE #28: feat: schema helpers + traverse (get_labels, get_edge_types, traverse)
## Summary

- `get_labels() → list[LabelInfo]` — list all node labels in schema
- `get_edge_types() → list[EdgeTypeInfo]` — list all edge types in schema
- `traverse(start_node_id, edge_type, direction, max_depth) → TraverseResult` — graph traversal RPC
- `LabelInfo`, `EdgeTypeInfo`, `PropertyDefinitionInfo`, `TraverseResult` result types
- 15 mock-based unit tests, 5 integration tests (outbound/inbound traverse, label/edge-type listing)

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- \`get_labels() → list[LabelInfo]\` — list all node labels from SchemaService
- \`get_edge_types() → list[EdgeTypeInfo]\` — list all edge types from SchemaService
- \`traverse(start_node_id, edge_type, direction, max_depth) → TraverseResult\` — graph traversal via GraphService/Traverse RPC
- \`LabelInfo\`, \`EdgeTypeInfo\`, \`PropertyDefinitionInfo\`, \`TraverseResult\` result types, all exported from \`coordinode.__init__\`

## Test Plan

- 15 mock-based unit tests (\`tests/unit/test_schema_crud.py\`, no server required)
- 5 integration tests (\`tests/integration/test_sdk.py\`): label listing, edge type listing, outbound traverse, property definitions
- Inbound traversal test marked \`@xfail(strict=False)\` — CoordiNode server does not yet return results for inbound direction. \`strict=False\` is intentional: XPASS (server gains inbound support) is good news, not a CI failure. The XPASS marker in pytest output is the signal to remove the decorator.

Closes #28

### CODE REVIEW

File: coordinode/coordinode/client.py [hunk 1]
```diff
@@ -303,6 +358,61 @@ async def get_schema_text(self) -> str:
...
+        )
+
+        _direction_map = {
+            "outbound": TraversalDirection.TRAVERSAL_DIRECTION_OUTBOUND,
+            "inbound": TraversalDirection.TRAVERSAL_DIRECTION_INBOUND,
+            "both": TraversalDirection.TRAVERSAL_DIRECTION_BOTH,
+        }
+        direction_value = _direction_map.get(
+            direction.lower(), TraversalDirection.TRAVERSAL_DIRECTION_OUTBOUND
+        )
```
[the remark this answers is not in the document]
Comment (polaz) [reply coordinode/coordinode/client.py]: @coderabbitai Fixed — now raises `ValueError` with message listing allowed values instead of silently defaulting.

File: tests/integration/test_sdk.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 import pytest
 
-from coordinode import AsyncCoordinodeClient, CoordinodeClient
+from coordinode import AsyncCoordinodeClient, CoordinodeClient, LabelInfo, EdgeTypeInfo, TraverseResult
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py]: @coderabbitai Fixed — imports sorted alphabetically via `ruff check --fix`.

File: tests/integration/test_sdk.py [hunk 3]
```diff
@@ -208,6 +208,98 @@ def test_get_schema_text(client):
...
+def test_get_labels_returns_list(client):
+    """get_labels() returns a non-empty list of LabelInfo after data is present."""
+    tag = uid()
+    client.cypher("CREATE (n:GetLabelsTest {tag: $tag})", params={"tag": tag})
+    try:
+        labels = client.get_labels()
+        assert isinstance(labels, list)
+        assert len(labels) > 0
+        assert all(isinstance(l, LabelInfo) for l in labels)
+        names = [l.name for l in labels]
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py]: @coderabbitai Fixed — renamed `l` → `lbl` in all three occurrences.

File: tests/integration/test_sdk.py [hunk 4]
```diff
@@ -208,6 +208,98 @@ def test_get_schema_text(client):
...
+    client.cypher(
+        "CREATE (a:TraverseRPC {tag: $tag, role: 'hub'})"
+        "-[:TRAVERSE_TEST]->(b:TraverseRPC {tag: $tag, role: 'leaf1'})",
+        params={"tag": tag},
+    )
+    rows = client.cypher(
+        "MATCH (a:TraverseRPC {tag: $tag, role: 'hub'}) RETURN a AS node_id",
+        params={"tag": tag},
+    )
+    try:
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py]: @coderabbitai Fixed — moved the ID-lookup cypher calls inside the `try` block so `finally` cleanup always runs.

File: tests/unit/test_schema_crud.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+"""
+
+from coordinode.client import (
+    EdgeResult,
+    EdgeTypeInfo,
+    LabelInfo,
+    NodeResult,
+    PropertyDefinitionInfo,
+    TraverseResult,
+)
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/test_schema_crud.py:20]: @coderabbitai Fixed — import block reformatted via `ruff check --fix`.

File: tests/integration/test_sdk.py [hunk 6]
```diff
@@ -208,6 +208,100 @@ def test_get_schema_text(client):
...
+    """LabelInfo.properties is a list (may be empty for schema-free labels)."""
+    client.cypher("MERGE (n:PropLabel {name: 'probe'})")
+    try:
+        labels = client.get_labels()
+        found = next((lbl for lbl in labels if lbl.name == "PropLabel"), None)
+        assert found is not None, "PropLabel not returned by get_labels()"
+        assert isinstance(found.properties, list)
+    finally:
+        client.cypher("MATCH (n:PropLabel {name: 'probe'}) DELETE n")
+
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py]: @coderabbitai Fixed — now uses `uid()` tag + `DETACH DELETE` for per-test isolation.

File: tests/integration/test_sdk.py [hunk 7]
```diff
@@ -208,6 +208,100 @@ def test_get_schema_text(client):
...
+        assert isinstance(result, TraverseResult)
+        assert len(result.nodes) >= 1, "traverse() returned no neighbour nodes"
+    finally:
+        client.cypher("MATCH (n:TraverseRPC {tag: $tag}) DETACH DELETE n", params={"tag": tag})
+
+
+@pytest.mark.skip(
+    reason="CoordiNode Traverse RPC does not yet support inbound direction — server returns empty result set"
+)
+def test_traverse_inbound_direction(client):
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py]: @coderabbitai Fixed — replaced `@skip` with `@xfail(strict=False)`.

File: tests/integration/test_sdk.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py]: @coderabbitai Fixed — added `if max_depth < 1: raise ValueError(...)` guard before `TraverseRequest` construction.

File: coordinode/coordinode/client.py [hunk 8]
```diff
@@ -303,6 +351,62 @@ async def get_schema_text(self) -> str:
...
+        if key not in _direction_map:
+            raise ValueError(f"Invalid direction {direction!r}. Must be one of: 'outbound', 'inbound', 'both'.")
+        direction_value = _direction_map[key]
+
+        req = TraverseRequest(
+            start_node_id=start_node_id,
+            edge_type=edge_type,
+            direction=direction_value,
+            max_depth=max_depth,
+        )
```
[the remark this answers is not in the document]
Comment (polaz) [reply coordinode/coordinode/client.py:416]: @coderabbitai Fixed — added `if max_depth < 1: raise ValueError(...)` guard before `TraverseRequest` construction.

File: tests/integration/test_sdk.py [hunk 9]
```diff
@@ -208,6 +208,102 @@ def test_get_schema_text(client):
...
+
+def test_get_labels_has_property_definitions(client):
+    """LabelInfo.properties is a list (may be empty for schema-free labels)."""
+    tag = uid()
+    client.cypher("CREATE (n:PropLabel {tag: $tag})", params={"tag": tag})
+    try:
+        labels = client.get_labels()
+        found = next((lbl for lbl in labels if lbl.name == "PropLabel"), None)
+        assert found is not None, "PropLabel not returned by get_labels()"
+        assert isinstance(found.properties, list)
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py:241]: @coderabbitai Won't strengthen assertion — CoordiNode is schema-free; `get_labels()` may return empty `properties` list even when nodes have properties. The docstring note 'may be empty for schema-free labels' is intentional. Added inline note to the test explaining this.

File: tests/integration/test_sdk.py [hunk 10]
```diff
@@ -208,6 +208,104 @@ def test_get_schema_text(client):
...
+    client.cypher("CREATE (n:PropLabel {tag: $tag})", params={"tag": tag})
+    try:
+        labels = client.get_labels()
+        found = next((lbl for lbl in labels if lbl.name == "PropLabel"), None)
+        assert found is not None, "PropLabel not returned by get_labels()"
+        # Intentionally only check the type — CoordiNode is schema-free and may return
+        # an empty properties list even when the node was created with properties.
+        assert isinstance(found.properties, list)
+    finally:
+        client.cypher("MATCH (n:PropLabel {tag: $tag}) DETACH DELETE n", params={"tag": tag})
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py]: Fixed — both tests now generate unique label and edge type names per run (e.g. `GetLabelsTest{uid()}` and `GET_EDGE_TYPE_TEST_{uid()}`) so stale schema entries from previous test runs cannot cause spurious assertion hits.

File: tests/integration/test_sdk.py [hunk 11]
```diff
@@ -208,6 +208,104 @@ def test_get_schema_text(client):
...
+        assert isinstance(result, TraverseResult)
+        assert len(result.nodes) >= 1, "traverse() returned no neighbour nodes"
+    finally:
+        client.cypher("MATCH (n:TraverseRPC {tag: $tag}) DETACH DELETE n", params={"tag": tag})
+
+
+@pytest.mark.xfail(
+    strict=True,
+    reason="CoordiNode Traverse RPC does not yet support inbound direction — server returns empty result set",
+)
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py:307]: Fixed — reverted to `strict=False`. The PR description already described `strict=False` semantics; the code is now aligned.

File: tests/unit/test_schema_crud.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+class TestEdgeTypeInfo:
+    def test_basic_fields(self):
+        et = EdgeTypeInfo(_FakeEdgeType("KNOWS", version=1))
+        assert et.name == "KNOWS"
+        assert et.version == 1
+        assert et.properties == []
+
+    def test_properties_are_wrapped(self):
+        props = [_FakePropDef("since", 6)]  # 6 = TIMESTAMP
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/test_schema_crud.py]: Fixed — introduced `PROPERTY_TYPE_TIMESTAMP = 6` class constant on `TestEdgeTypeInfo` and replaced the magic number, removing the inline comment that SonarCloud flagged as commented-out code.

File: tests/unit/test_schema_crud.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+    def test_edges_are_wrapped_as_edge_results(self):
+        edges = [_FakeEdge(10, "KNOWS", source=1, target=2)]
+        result = TraverseResult(_FakeTraverseResponse(edges=edges))
+        assert len(result.edges) == 1
+        assert isinstance(result.edges[0], EdgeResult)
+        assert result.edges[0].source_id == 1
+        assert result.edges[0].target_id == 2
+        assert result.edges[0].type == "KNOWS"
+
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/unit/test_schema_crud.py:180]: @coderabbitai Fixed — added `assert result.edges[0].id == 10` to verify the `edge_id → EdgeResult.id` mapping, consistent with the existing node id assertion pattern.

File: tests/integration/test_sdk.py [hunk 14]
```diff
@@ -208,6 +208,128 @@ def test_get_schema_text(client):
...
+        client.cypher("MATCH (n:TraverseRPC {tag: $tag}) DETACH DELETE n", params={"tag": tag})
+
+
+@pytest.mark.xfail(
+    strict=False,
+    # strict=False: XPASS is good news (server gained inbound support), not an error.
+    # strict=True would break CI exactly when the server improves, which is undesirable.
+    # The XPASS report in pytest output is the signal to remove this marker.
+    reason="CoordiNode Traverse RPC does not yet support inbound direction — server returns empty result set",
+)
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/test_sdk.py:307]: Fixed — added `raises=AssertionError` to narrow xfail to the known failure mode (empty result set → assertion fails). Unexpected errors such as gRPC `RpcError` will still propagate as CI failures.

File: coordinode/coordinode/client.py [hunk 15]
```diff
@@ -303,6 +351,68 @@ async def get_schema_text(self) -> str:
...
+            :class:`TraverseResult` with ``nodes`` and ``edges`` lists.
+        """
+        # Validate pure string/int inputs before importing proto stubs — ensures ValueError
+        # is raised even when proto stubs have not been generated yet.
+        _valid_directions = {"outbound", "inbound", "both"}
+        key = direction.lower()
+        if key not in _valid_directions:
+            raise ValueError(f"Invalid direction {direction!r}. Must be one of: 'outbound', 'inbound', 'both'.")
+        if max_depth < 1:
+            raise ValueError(f"max_depth must be >= 1, got {max_depth!r}.")
```
[the remark this answers is not in the document]
Comment (polaz) [reply coordinode/coordinode/client.py]: @coderabbitai Fixed — added `isinstance(direction, str)` before `.lower()` and `isinstance(max_depth, int) and not isinstance(max_depth, bool)` before bounds check. All three edge cases (None direction, str max_depth, bool max_depth) now raise `ValueError`. Three new unit tests added.
